### DISCUSSION TITLE
Official Tesla Third-Party Apps integration

### DISCUSSION THREAD

**Discussion by Niek** [teslamate-org/teslamate#3268] [General]
It's a bit preliminary, but Tesla is preparing to offer Third-Party Apps integration:
https://www.notateslaapp.com/software-updates/upcoming-features/id/1497/tesla-to-add-official-support-for-third-party-services-and-apps

Depending on what data will be available, if it's going to be free API access, etc - this could be interesting for Teslamate.

**Comment by LelandSindt** [comment]
This came across my news feed last night https://github.com/teslamotors/fleet-telemetry --- I can't quite tell if the two are related.

**Reply by Niek** [reply]
Great find! That's definitely related. I wonder if individuals can sign up for this, or only a limited set of service providers (see also: https://github.com/teslamotors/fleet-telemetry/issues/41).

If this is a self-service kind of integration, it should be possible to write a [custom dispatcher](https://github.com/teslamotors/fleet-telemetry/blob/main/telemetry/producer.go#L10-L19) for Teslamate.

**Comment by nelox** [comment]
As this story indicates, Tesla may be close to enabling Third-Party Apps integration:

> _A link to manage third party apps has been added to the Security and Privacy section of the app. A link to Manage Third Party Apps has been available on the web for about a month now, but given that it has now been added to the app, Tesla may be close to enabling this feature._

https://www.notateslaapp.com/tesla-app-updates/version/4.23.5/release-notes

**Comment by hebrandt** [comment]
Rest Api has been deprecated on 17.11.2023 (https://developer.tesla.com/docs/fleet-api#fleet-telemetry)

**Reply by okazakov** [reply]
The old REST API will stop working in January 2024. As documented in Tesla's developer doco linked above:
_"January 2024 | All vehicles* will require Tesla Vehicle Command protocol. The REST API will be fully deprecated"_

Is there a plan to transition to the new Tesla Vehicle SDK? Since we're in mid December now, NOW is the time to be doing something about it...

Cheers.

**Reply by brianmay** [reply]
People are looking into this. But this may not be possible for open source software such as TeslaMate. e.g. the new API requires creating a private/public key pair, and then registering the public key with Tesla. Tesla has a number of hurdles that must be overcome in order to do this. e.g. you need to be have a legal business. See https://github.com/teslamate-org/teslamate/discussions/3527

And Tesla also dropping hints that this is going to be a paid for API. Based on what happened with the Twitter API, I imagine that fees are going to be for the high end fleet type operations, and probably out of reach for the average TeslaMate user.

I guess we have to wait and see what happens. I think that Tesla doesn't really care about (or even acknowledge) the open source community.

The lack of good solid information on what is happening (e.g. they aren't even providing the source proto files for the new API), and when it is happening, means that any time spent on this is taking a risk, as we could easily suddenly find there is no API we can use anymore.

I imagine that any solution, if there is one, may not come until after the shut down of the existing API. Simply because of the lack of information from Tesla. After the shutdown we can see what is working, what is not working, and work from that.

**Reply by CoMPaTech** [reply]
The way I understand the information on the Tesla page is that 'sometime in Januari' (hopefully Elon-time) the old API will be shut down. New users (i.e. new cars) starting November could not get access on the old API. So anyone still on very early S and X builds + everyone registering before that mid-November is still good to go ... that is, until they shut it.

**Reply by SylvainGa** [reply]
End of month is approaching and my Teslamate is still getting data from my car. Let's not wake Elon lol.

**Reply by weljajoh** [reply]
Well, judging by the error from one of my own scripts that failed as of this morning, my car has been migrated to the new API.

`2024-01-23 07:28:05 vehicle_list() HTTPError: 412 Client Error: Endpoint is only available on fleetapi. Visit https://developer.tesla.com/docs for more info for url: https://owner-api.teslamotors.com/api/1/vehicles`

**Reply by wooter** [reply]
I got a notification from TeslaFi this morning that the new API is active and my car wasn't receiving any updates from the old API. I had to re-authenticate my car on the new API. However, as of this morning, Teslamate still receives information of my car on the old API.

**Reply by brianmay** [reply]
See #3629 . In short, I think an existing running instance will work fine, but if you restart it for any reason, it may not work anymore.

**Reply by brianmay** [reply]
That message seems to imply that the rest fleet API still works. Don't know if we could use that or not. Guessing the tokens we have won't work for the fleet API.

**Comment by brianmay** [comment]
In my logs I have the dump of the `TeslaApi.Vehicle` object.

```
Jan 11 11:20:20 iot2 teslamate[1254]: %TeslaApi.Vehicle{
Jan 11 11:20:20 iot2 teslamate[1254]:   id: <censored>,
Jan 11 11:20:20 iot2 teslamate[1254]:   vehicle_id: <censored>,
Jan 11 11:20:20 iot2 teslamate[1254]:   vin: <censored>,
Jan 11 11:20:20 iot2 teslamate[1254]:   tokens: <censored>,
Jan 11 11:20:20 iot2 teslamate[1254]:   state: "unknown",
Jan 11 11:20:20 iot2 teslamate[1254]:   option_codes: [""],
Jan 11 11:20:20 iot2 teslamate[1254]:   in_service: false,
Jan 11 11:20:20 iot2 teslamate[1254]:   display_name: nil,
Jan 11 11:20:20 iot2 teslamate[1254]:   color: nil,
Jan 11 11:20:20 iot2 teslamate[1254]:   calendar_enabled: true,
Jan 11 11:20:20 iot2 teslamate[1254]:   backseat_token: nil,
Jan 11 11:20:20 iot2 teslamate[1254]:   backseat_token_updated_at: nil,
Jan 11 11:20:20 iot2 teslamate[1254]:   api_version: 71,
[...]
```

Would suggest getting a copy of these values if you have it. Might be possible to implement some sort of work around if you have the `id`, `vehicle_id`, `vin` and `display_name` fields (probably should check the source code to see which values we use).
